Adaptive walnuts

bob-carpenter opened the pull request:
I think everything's working as intended now in `AdaptiveWalnuts`.  

I'd really appreciate if someone can check whether I got the move semantics right.  Any other tips appreciated, of course.  

I added Doxygen doc and a `doc` target to `cmake`.  It builds into `doc/html/index.html`.  Checking the `Cmakelists.txt` file would also be helpful for me.  GPT had a bunch of suggestions for improving it as it helped me write the doc.  But as you saw last time, it's not so great at `Cmake` config, so some human eyes would be great.  We can talk about refactoring the `Cmake` files later.

I set up a pattern for adaptive WALNUTS using a class `AdaptiveWalnuts` that provides a functor-based generator for the nest warmup state, then a function that converts the adaptive sampler into a fixed WALNUTS sampler.  You can see an example use in `examples/test.cpp`.

Still no tests other than the for the base estimators, which are not stochastic and the end-to-end eyeball tests in `examples/test.cpp`.

I still haven't added cross-refs to the docs.  I also haven't documented the base `NUTS` class.

I tried to provide hints about the algorithm in the docs, but I think we're going to need an exterior algorithm definition as I'm creating in Overleaf and will post on *arXiv* (it's a more computationally-focused alternative to our theory paper).

bob-carpenter commented on walnuts_cpp/include/walnuts/walnuts.hpp:
```diff
@@ -264,5 +334,116 @@ void walnuts(Generator &generator, const F &logp_grad_fun,
+ * @relates WalnutsSampler
+ */
+template <class F, typename S, class RNG>
+WalnutsSampler(RNG&, F&, const Vec<S>&, const Vec<S>&, S, Integer, S)
```
It's not necessary for the use from `AdaptiveWalnuts`, which does specify all types. 

It's useful for a client to call the constructor without having to fill in tricky types like `F`, for example, this usage from `test.cpp`:

```cpp
  nuts::WalnutsSampler sample(rand, normal_logp_grad, theta_init,
			      inv_mass, macro_step_size, max_nuts_depth,
			      log_max_error);
```

bob-carpenter replied:
Nice.  I got rid of all the deduction guides.  I had to add a `Wno-ctad-maybe-unsupported` flag in the Cmake file.  Should be fine in any compiler supporting all of C++17.

WardBrian commented on walnuts_cpp/include/walnuts/adaptive_walnuts.hpp:
```diff
@@ -0,0 +1,411 @@
+  AdaptiveWalnuts(RNG& rng,
+                  F& logp_grad,
+                  const Vec<S>& theta_init,
+                  MassAdaptConfig<S>&& mass_cfg,
```
The nice thing about perfect forwarding is then the call site can decide if they want theta again afterwards

bob-carpenter commented on walnuts_cpp/include/walnuts/walnuts.hpp:
```diff
@@ -241,17 +304,24 @@ Vec<S> transition_w(Random<S, Generator> &rng, const F &logp_grad_fun,
+class NoOpHandler {
+ public:
+  template <typename T>
+  void operator()(const T&) const noexcept { }
```
Forcing inlining requires a bunch of compiler-specific branching to accomplish, which I'd rather only do if necessary. 

From what I can tell, it looks like at `-O2`, modern compilers will inline everything that won't increase code size, which means this will get inlined, even if not declared so.  I added the declaration since it won't hurt anything and may help readers.

WardBrian replied:
Since we're using modern-ish standards you can get the major compilers without any `#ifdefs` using attributes. namely `[[gnu::always_inline]] [[msvc::forceinline]]` will hit get, clang, and MSVC

We do this one place in [bridgestan](https://github.com/roualdes/bridgestan/blob/bb01fd8234212b934e0dc81a2ead72c6ac15ebf6/src/util.hpp#L38)

bob-carpenter replied:
Thanks.  That's not necessary here.  Apparently all the -O2 and above compilers will inline anything that reduces code size.  A no-op reduces code size, so we should be good to go without any declarations here.

bob-carpenter commented:
I'm not sure it's worth maintaining the NUTS implementation here.  If it is, it's probably also worth adding a similar adaptation for it and converting it over to the same functor-style implementation as for walnuts.  I'd estimate those updates plus documentation would require 10 to 20 hours.

The question is for an eventual wrapper, do we also want to supply a NUTS implementation with this new continuous adaptation?  I'm just not sure it's ever going to make sense to use NUTS with WALNUTS available.  But it would be good for comparisons.

WardBrian commented:
@bob-carpenter I've just pushed a commit to update the test_stan program to match the behavior of the updated test program

WardBrian commented on walnuts_cpp/CMakeLists.txt:
```diff
@@ -143,3 +147,29 @@ if (CLANG_FORMAT_BIN)
+     "${CMAKE_CURRENT_SOURCE_DIR}/include/walnuts/*.hpp")
+
+# Doxygen configuration overrides
+set(DOXYGEN_OUTPUT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/doc")
```
I don't think we should have this output into the source directory, personally

bob-carpenter replied:
To try to avoid all the confusion we've been having, I'm trying to stick to "client" for users of the code and "developers" for people who are building it.  When you say "user's" here, do you mean developers?  I don't think we should expect clients to download Doxygen and CMake to build our docs, so I can't imagine you mean clients.

WardBrian replied:
I don't believe I have used the word clients in this thread. Really all I am saying is I think we should delete this `set` line -- the default cmake picks for us is more conventional.

If a user wants to build our docs themselves, they certainly will need cmake and doxygen, but I'm not sure that is relevant to this discussion.

bob-carpenter commented:
I brought up the term "client" to try to distinguish regular users from devs because I was getting confused about who we were talking about.  I find the naming of "`CURRENT_SOURCE_DIR`" very confusing when it's also the superdirectory of the build directory!

I deleted the offending line of CMake config:

```cmake
set(DOXYGEN_OUTPUT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/doc")
```

WardBrian commented:
As an aside: It would be nice if the Walnuts class let you query what the e.g. macro stepsize is, so that you can access it after it has been adapted by AdaptiveWalnuts

bob-carpenter commented:
Great point.  I'll add that to the WALNUTS class.  It's been a huge pain not being able to access this kind of thing easily through the Stan interfaces (you can see the hack of putting them in as comments in the csv output, which has led to no end of brittleness and inefficiency in CSV parsing).

WardBrian commented:
Adaptive walnuts seems to fail entirely if you initialize a standard normal at 0.

Changing line 174 of `test.cpp` to just assign `theta_init(i) = 0.0;` yields

```
TEST ADAPTIVE WALNUTS
dim 0: mean = 0, stddev = 0
dim 1: mean = 0, stddev = 0
dim 2: mean = 0, stddev = 0
dim 3: mean = 0, stddev = 0
... elided 4 rows ...
dim 7: mean = 0, stddev = 0
dim 8: mean = 0, stddev = 0
dim 9: mean = 0, stddev = 0

Macro step size = 1
Max error = 0.1
Inverse mass matrix = inf inf inf inf inf inf inf inf inf inf
```

bob-carpenter commented:
I retooled the tests so I'm using the same timing information.  Were you resetting the global counters?

Here's what I'm getting now for a 250D normal with 0.1 target error with NUTS using the same step size as the macro step size tuned by WALNUTS.  This is the one with standard deviation = 1 + dimension.   I'm running adaptive WALNUTS for `N` warmup iterations *and* `N` sampling iterations, so we'd expect a factor of 2 or so.  In a realistic setting, we'd also be running NUTS adaptation for twice as long as we run NUTS.

```
TEST NUTS;  D = 250;  N = 2000;  step_size = 0.425;  max_depth = 10
     logp_grad calls: 2045952
          total time: 3.46307s
      logp_grad time: 1.91207s
  logp_grad fraction: 0.552133

TEST WALNUTS;  D = 250;  N = 2000;  macro_step_size = 0.425;  max_nuts_depth = 10;  max_error = 0.1
     logp_grad calls: 2033645
          total time: 3.90068s
      logp_grad time: 3.77436s
  logp_grad fraction: 0.967617

TEST ADAPTIVE WALNUTS;  D = 250;  N = 2000; step_size_init = 1; max_nuts_depth = 10; max_error = 0.1
     logp_grad calls: 99126
          total time: 0.154715s
      logp_grad time: 3.86495s
  logp_grad fraction: 24.9811
```

Here, NUTS and WALNUTS are similar, but adaptive WALNUTS crushes them because it can adapt to the ill-conditioned problem.

And here's what I get with a 250D standard normal target at max error of 0.1.  In this case, both NUTS and WALNUTS are using the perfect mass matrix, but the one adapted by adaptive WALNUTS has about 10% error.  So we pay a bit of a price for having to adapt rather than having the perfect mass matrix.

```
TEST NUTS;  D = 250;  N = 2000;  step_size = 0.425;  max_depth = 10
     logp_grad calls: 16000
          total time: 0.0473609s
      logp_grad time: 0.00358236s
  logp_grad fraction: 0.0756397

dim 0: mean = 0.0019935, stddev = 0.998872
dim 1: mean = -0.0047783, stddev = 0.969022
dim 2: mean = -0.00161321, stddev = 0.978471
dim 3: mean = 0.0157197, stddev = 1.01458
... elided 244 rows ...
dim 247: mean = -0.00464528, stddev = 0.983911
dim 248: mean = 0.0351389, stddev = 1.02854
dim 249: mean = 0.0242598, stddev = 1.00102

TEST WALNUTS;  D = 250;  N = 2000;  macro_step_size = 0.425;  max_nuts_depth = 10;  max_error = 0.1
     logp_grad calls: 49518
          total time: 0.0440712s
      logp_grad time: 0.00928165s
  logp_grad fraction: 0.210606

TEST ADAPTIVE WALNUTS;  D = 250;  N = 2000; step_size_init = 1; max_nuts_depth = 10; max_error = 0.1
     logp_grad calls: 103496
          total time: 0.0739357s
      logp_grad time: 0.0190264s
  logp_grad fraction: 0.257337
```

Now WALNUTS is about 3 times as many calls, and adaptive WALNUTS about 6 times as many (for twice as many iterations).  

Here's the results for setting the max error to be 0.5, which is about what I think @torekleppe used in the paper's evaluations.

```
TEST NUTS;  D = 250;  N = 2000;  step_size = 0.425;  max_depth = 10
     logp_grad calls: 16000
          total time: 0.0477692s
      logp_grad time: 0.00368012s
  logp_grad fraction: 0.0770397

dim 0: mean = 0.0019935, stddev = 0.998872
dim 1: mean = -0.0047783, stddev = 0.969022
dim 2: mean = -0.00161321, stddev = 0.978471
dim 3: mean = 0.0157197, stddev = 1.01458
... elided 244 rows ...
dim 247: mean = -0.00464528, stddev = 0.983911
dim 248: mean = 0.0351389, stddev = 1.02854
dim 249: mean = 0.0242598, stddev = 1.00102

TEST WALNUTS;  D = 250;  N = 2000;  macro_step_size = 0.425;  max_nuts_depth = 10;  max_error = 0.5
     logp_grad calls: 18764
          total time: 0.0324169s
      logp_grad time: 0.00592513s
  logp_grad fraction: 0.182779

dim 0: mean = 0.0367216, stddev = 0.97273
dim 1: mean = 0.00216988, stddev = 0.98402
dim 2: mean = 0.00615332, stddev = 0.998804
dim 3: mean = -0.00582768, stddev = 0.990048
... elided 244 rows ...
dim 247: mean = -0.0321928, stddev = 0.990814
dim 248: mean = 0.0351662, stddev = 1.03525
dim 249: mean = 0.0313886, stddev = 1.03945

TEST ADAPTIVE WALNUTS;  D = 250;  N = 2000; step_size_init = 1; max_nuts_depth = 10; max_error = 0.5
     logp_grad calls: 40367
          total time: 0.0533893s
      logp_grad time: 0.00978974s
  logp_grad fraction: 0.183365
```

Now they're pretty much in line, as we'd expect.  I just had a pretty conservative setting at 0.1 max error.  

Could you double check my instrumentation code in `examples/test.cpp`?  It's using globals and functions.  I think it's OK because the answers are sensible.

bob-carpenter commented:
I found the bug with initializing at the mode---the gradients are zero, so the precisions get initialized at zero and that means the variances get initialized at +infinity.  I put in some additive smoothing, which fixes the problem, but it's another tuning parameter to think about.

I'm still seeing WALNUTS doing a lot more work in some situations than NUTS, so I still need to do the big evals. 

But I think at this point, I'll just merge this and then go on to further refinements before the entire project winds up in one PR.

WardBrian commented:
It seems like nutpie's strategy is just ignore all-zero or all-infinite values in the mass matrix updates, optionally replacing them with a fill value: https://github.com/pymc-devs/nuts-rs/blob/bbfc69a2f61b06c5916c661d8cdcd7704710bd88/src/cpu_math.rs#L370-L375

Anyway, using 0.5 for the max error does seem to get me within a factor of 1.5 of NUTS' logp calls

WardBrian commented:
> TEST ADAPTIVE WALNUTS;  D = 250;  N = 2000; step_size_init = 1; max_nuts_depth = 10; max_error = 0.1
>      logp_grad calls: 99126
>           total time: 0.154715s
>       logp_grad time: 3.86495s
>   logp_grad fraction: 24.9811

Something was clearly wrong at this point, because the fraction is greater than 1. I think you're not resetting total_time anywhere.

bob-carpenter commented:
Good catch---I was forgetting to reset the gradient timer.

With max error equal to 10.0, the precision test never fails, and WALNUTS devolves to NUTS.  In the test cases, NUTS and WALNUTS take the same amount of time and number of gradients.  As that number goes down, WALNUTS will be more precise and take more time.

@torekleppe has some ideas on how to tune the max error, which he used in the Python code for the paper.  I was really hoping there would be a reasonable universal choice for it like there is for the acceptance rate in NUTS, for which values between 0.6 and 0.8 are usually fine.

WardBrian commented on walnuts_cpp/include/walnuts/adaptive_walnuts.hpp:
```diff
@@ -0,0 +1,399 @@
+        rand_,
+        logp_grad_,
+        theta_,
+        mass_adapt_var_.variance(),
```
Shouldn't the mass matrix for the new sampler also use `mass_adapt_prec_`?

bob-carpenter replied:
Yes, it should.  I have no idea how I let that slip through!

bob-carpenter replied:
I pushed a fix that encapsulates the mass matrix estimator to avoid cutting-and-pasting and then added the extra config to the mass estimator config. The sampling code is much cleaner and the mass matrix estimators are better.

bob-carpenter commented:
I think this is in good shape to move over to a new repo at this point.  @WardBrian---can you do that or should I give it a shot?

* I've double checked all the doc and renamed some variables for clarity.  
* I formatted everything with our clang format.  
* I rewrote the `README.md` file to group all the builds together and listed all the dependencies and licensing. 
* I broke out the mass matrix estimator in a class to encapsulate since I need to use some of the functions multiple times and I need to hold onto a bunch of sub-estimators.  
* I changed the stopping condition and within-tolerance check to only check tolerances at the macro steps as in the paper
     - this adds a bunch conditionals and scalar accumulators in the tight loops
     - we only use the positions at the macro steps, so we don't need accuracy on the micro steps and thus get looser tolerances
     - we lose the ability to fail fast with the every step accumulator (but error tends to grow, so maybe not such a big deal?)
* I fixed the bug where I wasn't using the right estimator for the adapted WALNUTS (thanks again for catching that @WardBrian)
* Reworked the timers for examples into functions
* Conditioned the initialization of mass matrix estimators for dealing with zero gradients (may want to think longer term about better conditioning here, but at least it solves the problem of initializing at the mode [which we do *not* ever want to do because it's a bottleneck for HMC sampling because you're moving from a very high density state to something smaller density and thus have to take small steps to get decent acceptance])

I should probably move the file `examples/test.cpp` to something with an "example" in the name and then have the target not be `./test_nuts`, which I find confusing from a `test.cpp` file in this simple case.

WardBrian commented:
I've started moving this over to https://github.com/flatironinstitute/walnuts, which we can make public as soon as you confirm it looks like you would have expected

bob-carpenter commented:
This looks great.  The only thing it's missing is the issue and PR history, but I think that's fine.  If I was the only one who'd done any work, I would've suggested just starting over from a clean repo.
